Add elem_offset and related methods

wr7 opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r​? <reviewer name>
-->
<!-- homu-ignore:end -->

Implementation of #126769

wr7 committed:
Add elem_offset and related methods

tgross35 commented on library/core/src/slice/mod.rs:
```diff
@@ -4540,6 +4540,120 @@ impl<T> [T] {
         unsafe { Ok(self.get_many_unchecked_mut(indices)) }
     }
+
+    /// Finds the index that an element points to within the slice.
```
"element points to" doesn't sound right since the elements aren't necessarily pointers. Maybe

```suggestion
    /// Locate a pointed-to item within a slice from its reference.
```

wr7 replied:
Good catch. I'm going to go with `Returns the index that an element reference points to` for now.

tgross35 commented on library/core/src/slice/mod.rs:
```diff
@@ -4540,6 +4540,120 @@ impl<T> [T] {
+    ///
+    /// Note that this uses pointer arithmetic and **does not compare elements**.
+    /// To find the index of an element via comparison, use
+    /// [`.iter().position()`](crate::iter::Iterator::position) instead
```
```suggestion
    /// [`.iter().position()`](crate::iter::Iterator::position) instead.
```

tgross35 commented on library/core/src/slice/mod.rs:
```diff
@@ -4540,6 +4540,120 @@ impl<T> [T] {
+        let elem_start = element as *const T as usize;
+
+        let byte_offset = elem_start.wrapping_sub(self_start);
+        let offset = byte_offset / mem::size_of::<T>();
+
+        if byte_offset % mem::size_of::<T>() != 0 {
+            return None;
+        }
+
+        if offset < self.len() { Some(offset) } else { None }
```
```suggestion
        if byte_offset % mem::size_of::<T>() != 0 {
            return None;
        }
        
        let offset = byte_offset / mem::size_of::<T>();
        if offset < self.len() { Some(offset) } else { None }
```

No need to calculate the offset if you might return early (the optimizer will get this too of course)

wr7 commented:
I went ahead and applied the changes suggested by tgross35. I also removed "within the slice" because it's unnecessarily verbose.

Amanieu commented on library/core/src/str/mod.rs:
```diff
@@ -2721,6 +2722,39 @@ impl str {
         EscapeUnicode { inner: self.chars().flat_map(CharEscapeUnicode) }
     }
+
+    /// Returns the range that a sub`str` points to.
```
```suggestion
    /// Returns the range that a substring points to.
```

Amanieu commented:
r=me once the comment is fixed

wr7 commented:
I fixed the comment, but my text editor auto-formated and changed some random stuff. I just rebased it back to normal.

This is my first time creating a PR to rust. Is this what you mean by `r=me`?

@bors r=Amanieu

bors commented:
@wr7: :key: Insufficient privileges: Not in reviewers

wr7 commented:
Or maybe this?

r? @Amanieu

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 557210c5c7f4045a2007583c741476a4a3aee1df has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu 557210c5c7f4045a2007583c741476a4a3aee1df -->
<!-- homu: {"type":"Approved","sha":"557210c5c7f4045a2007583c741476a4a3aee1df","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
